savant/base/pyfunc.py

```diff
@@ -70,6 +71,34 @@ def process_buffer(self, buffer: Gst.Buffer):
         :param buffer: Gstreamer buffer.
         """
 
+    def get_queues(self) -> List[Gst.Element]:
+        """Get upstream queues.
+
+        Can be used to check the current queue length for bandwidth management.
+
+        E.g.:
+        ```
+        for queue in self.get_queues():
+            queue_length = queue.get_property('current-level-buffers')
+            if queue_length > 10:
+                self.logger.warning('Queue %s is full!', queue.get_name())
+        ```
+        """
+        if self.gst_element is not None:
+            gst_pipeline = self.gst_element.get_parent()
+            return get_elements(
+                gst_pipeline,
+                factory='queue',
+                before_element=self.gst_element,
```

Comment:
@denisvmedyantsev Why this limitation is applied? Do not you think that it can be important to know not only queues before a particular element, but also after?

Reply:
From the issue description "we need to have an option to access the stats of all pyfuncs positioned before them", I thought here should be "queue" instead of "pyfunc" -> "we need to have an option to access the stats of all queues positioned before the pyfunc". No?

"The same, if we return queues, we need a concrete type for them." - I don't get it, please clarify.

Reply:
Now pyfunc has methods: 
- get_queues() returns all queues in the pipeline
- get_queue(name) returns queue by name (any queue, up or downstream the pyfunc)
- get_upstream_queue() returns the last (the nearest) queue before the pyfunc